pkg/icingaredis/telemetry/stats.go

```diff
@@ -0,0 +1,69 @@
+package telemetry
+
+import (
+	"context"
+	"github.com/go-redis/redis/v8"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/icingaredis"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/periodic"
+	"github.com/icinga/icingadb/pkg/utils"
+	"go.uber.org/zap"
+	"strconv"
+	"time"
+)
+
+// Stats periodically forward reported metrics to Redis for being monitored by Icinga 2.
+type Stats struct {
+	// Config & co. are to be increased by the T sync once for every T object synced.
+	Config, State, History, Overdue, HistoryCleanup com.Counter
+
+	ctx    context.Context
+	redis  *icingaredis.Client
+	logger *logging.Logger
+}
+
+// NewStats creates new Stats.
+func NewStats(ctx context.Context, redis *icingaredis.Client, logger *logging.Logger) *Stats {
+	s := &Stats{
+		ctx:    ctx,
+		redis:  redis,
+		logger: logger,
+	}
+
+	s.start()
+
+	return s
+}
+
+// start kicks off the actual work.
+func (s *Stats) start() {
+	counters := map[string]*com.Counter{
+		"sync_config":     &s.Config,
+		"sync_state":      &s.State,
+		"sync_history":    &s.History,
+		"sync_overdue":    &s.Overdue,
+		"cleanup_history": &s.HistoryCleanup,
+	}
+
+	periodic.Start(s.ctx, time.Second, func(_ periodic.Tick) {
+		var data []string
+		for kind, counter := range counters {
+			if cnt := counter.Reset(); cnt > 0 {
+				data = append(data, kind, strconv.FormatUint(cnt, 10))
+			}
+		}
+
+		if data != nil {
+			cmd := s.redis.XAdd(s.ctx, &redis.XAddArgs{
```

Comment:
Easier like I could avoid the surrounding if?

Reply:
Go ahead and write the code that calculates the rates from this data. Things might be easier there when you can just take the last N values from the stream, but could also work out this way, so let's get back to this when the corresponding C++ code exists.

Reply:
```py
for (id, metrics) in stream:
 for (name, v) in metrics:
  ringBuffers[name].InsertValue(getNixTs(id), v)
```

Reply:
`getNixTs(id)` would be the interesting part there. Timing isn't perfect and clocks can drift, so if you're simply rounding, you will get some seconds with no or two values. So you'd have to to consider the delta between the timestamps and round them to full seconds. Alternatively, you could to get the 1 minute average, sum up all values that are `> $last_ts - (1 minute + 0.5s)` (0.5s chosen by some eyeballing, there might be better values to ensure you get the values from exactly 60 seconds).

On the other hand, if you write every second, you can just take the last N values from the stream. Then check if there is a gap greater say 1.5s in the data. There typically shouldn't be one unless Icinga DB was restarted or otherwise gone, so discarding previous values should be fine there.

Reply:
> get the values from exactly 60 seconds

IYO shall I write stats once/s or /3s?

Reply:
What's the smallest interval the check should show a rate in the end? 1 minute? So the write interval should be quite a bit smaller than that and any realistic check_interval (otherwise your graphs might end up looking strange if you start to see these write intervals). So from that perspective, both 1s or 3s intervals will be fine. So I don't thing this would make a big difference apart from 1s being the nice number because the unit seconds was chosen that way. What would be the motivation for 3s? I mean this is only writing some data (probably way less than 1KB) to Redis, so I don't see running into any performance issues here.

tl;dr: 3s would be fine for everything I can imagine right now being implemented in the check, but 1s isn't so excessive that we shouldn't be able to afford just going with the nicer number.

Reply:
> get the values from exactly 60 seconds

I meant: I could easily s/3s/1s/. Shall I do this to make the numbers more exact?

Reply:
So to reduce the possible error from 1/20 to 1/60 (assuming all numbers are roughly the same magnitude, outliers would contribute more error)? Ideally, the code reading the data is written so that this difference wouldn't matter (apart from returning a new value every 1s vs. every 3s).